### ISSUE TITLE
Ergonomic testing of examples

### ISSUE THREAD

**Issue by azriel91** [assert-rs/assert_cmd#70]
Hiya, after the switch for [faster bin lookup](https://github.com/assert-rs/assert_cmd/commit/9379147429ff1eb8cb0766c696d1ae6141b66a33) (released in 0.11), writing code to assert on examples has become quite verbose, albeit workable (trade offs!).

<details>
<summary>

Code comparison (0.10 vs 0.11 with [`escargot`](https://github.com/crate-ci/escargot))

</summary>

Previously (0.10):

```rust
use std::process::Command;

use assert_cmd::{cargo::CargoError, prelude::*};

#[test]
fn example_01_draw_text() -> Result<(), CargoError> {
    Command::cargo_example("01_draw_text")?
        .env("APP_DIR", env!("CARGO_MANIFEST_DIR"))
        .args(&["--timeout", "0"])
        .output()
        .unwrap()
        .assert()
        .success();

    Ok(())
}
```

Currently (0.11):

```rust
use assert_cmd::{cmd::OutputError, prelude::*};
use escargot::CargoBuild;

#[test]
fn example_01_draw_text() -> Result<(), OutputError> {
    CargoBuild::new()
        .example("01_draw_text")
        .current_release()
        .current_target()
        .run()
        .expect("Failed to create `cargo` command")
        .command()
        .env("APP_DIR", env!("CARGO_MANIFEST_DIR"))
        .args(&["--timeout", "0"])
        .ok()?
        .assert()
        .success();

    Ok(())
}
```

</details>

#### Possible Solution

With the `--message-format` parameter, we get get `cargo build` to tell us the path to the example executable:

```bash
cargo build --examples -p application_ui --message-format json
```

```json
// .. 323 lines of other artifacts
{"reason":"compiler-artifact","package_id":"application_ui 0.9.0 (path+file:///D:/work/gitlab/azriel91/autexousious/crate/application_ui)","target":{"kind":["example"],"crate_types":["bin"],"name":"01_draw_text","src_path":"D:\\work\\gitlab\\azriel91\\autexousious\\crate\\application_ui\\examples/01_draw_text/main.rs","edition":"2018"},"profile":{"opt_level":"0","debuginfo":2,"debug_assertions":true,"overflow_checks":true,"test":false},"features":[],"filenames":["D:\\work\\gitlab\\azriel91\\autexousious\\target\\debug\\examples\\01_draw_text.exe"],"executable":"D:\\work\\gitlab\\azriel91\\autexousious\\target\\debug\\examples\\01_draw_text.exe","fresh":true}
```

The relevant bit is:

```json
"executable":"D:\\work\\gitlab\\azriel91\\autexousious\\target\\debug\\examples\\01_draw_text.exe"
```

I didn't figure out if `cargo-metadata` allows you to get similar information, but if we could get that information quickly (possibly within `assert_cmd`), and not require users to compile `cargo`, it's likely to be a nice solution.

**Comment by epage** [comment]
<details>
<summary>

- Simplified "0.10"s use of `assert`, update "0.11" example to match
- Added back in `current_target`, `current_release` because that is what `assert_cmd` was doing
- Removed "expect" from "0.11" to match "0.10".

</summary>

Previously (0.10):

```rust
use std::process::Command;

use assert_cmd::{cargo::CargoError, prelude::*};

#[test]
fn example_01_draw_text() -> Result<(), CargoError> {
    Command::cargo_example("01_draw_text")?
        .env("APP_DIR", env!("CARGO_MANIFEST_DIR"))
        .args(&["--timeout", "0"])
        .assert()
        .success();

    Ok(())
}
```

Currently (0.11):

```rust
use assert_cmd::{cmd::OutputError, prelude::*};
use escargot::CargoBuild;

#[test]
fn example_01_draw_text() -> Result<(), OutputError> {
    CargoBuild::new()
        .example("01_draw_text")
        .current_release()
        .current_target()
        .run()
        .command()
        .env("APP_DIR", env!("CARGO_MANIFEST_DIR"))
        .args(&["--timeout", "0"])
        .assert()
        .success(); // chain is pretty long now

    Ok(())
}
```

</details>

**Comment by epage** [comment]
> I didn't figure out if `cargo-metadata` allows you to get similar information, but if we could get that information quickly (possibly within `assert_cmd`), and not require users to compile `cargo`, it's likely to be a nice solution.

So to be clear, we are not linking against `cargo` but `assert_cmd` was using `escargot` to make calls out to `cargo` exactly how you describe which is a similar technique to `cargo_metadata`.

**Comment by epage** [comment]
So explanation (summarizing the issues linked to in that commit)

We've had multiple problems with hiding the use of `escargot`
- Overhead of calling `cargo` on the command line
- if we specify `current_target`, you slow down `cargo test` where `--target` isn't specified (typical developer workflow(.  If we leave out `current_target`, it reduces coverage in unexpected ways.

There have been a lot of cases where we've been recommending and people have had to drop down to calling `escargot` directly.

The new technique gives us some decent default at the cost of pushing more people to `escargot` which, I'm assuming, they should be using anyways instead of a hidden call to `escargot` with these hidden trade offs.

**Comment by epage** [comment]
To be clear, I am very interested in people's experience after this switch to try to help gauge any further changes.  This is the reason I've not released this as 1.0 yet.

After getting a little more background on the constraints, I'd be interested to know more of your thoughts.

**Comment by azriel91** [comment]
Hiya, after a bit of experimentation I think I understand some more:

* What things look like in 0.11 (annotated):

    ```rust
    use assert_cmd::assert::OutputAssertExt;
    use escargot::{error::CargoError, CargoBuild};

    // Error type is `CargoError`, so we can `run()?`
    // https://docs.rs/escargot/0.4.0/escargot/struct.CargoBuild.html#method.run
    #[test]
    fn example_01_draw_text() -> Result<(), CargoError> {
        CargoBuild::new()
            .example("01_draw_text")

            // Ensure that `--release` is specified if the test is compiled in release mode
            // If you miss this, then a `release` mode test will compile the example in
            // `debug` mode.
            .current_release()

            // May or may not be included -- see next bullet point
            .current_target()

            // `?` unwraps the `CargoResult`
            .run()?
            .command()
            .env("APP_DIR", env!("CARGO_MANIFEST_DIR"))
            .args(&["--timeout", "0"])
            .assert()
            .success();

        Ok(())
    }
    ```

* `current_target()`:

    - Specifying this means Cargo builds in the `target/<target>/<profile>` directory.

        Example: `target/x86_64-pc-windows-msvc/release`

    - Not specifying this means Cargo builds in the `target/<profile>` directory.

        Example: `target/release`

    The following table clarifies the implications of this when running `cargo test`:

    | &ndash;         | with `current_target()`  | without |
    | ---             | ---                      | ---     |
    | with `--target` | Cmpl: single, Cvrg: okay | Cmpl: multi, Cvrg: reduced |
    | without         | Cmpl: multi, Cvrg: okay  | Cmpl: single, Cvrg: reduced |

    Takeaway points:

    - Multiple compilations = slow; slow = bad.
    - Reduced coverage = inaccurate quality checks = bad.
    - Specifying `--target` during development is inconvenient and is easy to forget, and not all CI jobs specify it explicitly.

---

As for the switch experience, it was something like this:

1. There's a breaking change -- `Command::cargo_example` no longer exists. Let's check `CHANGELOG.md`.
2. It links to a commit, perhaps I'll check [`docs.rs`](https://docs.rs/assert_cmd/).
3. There's nothing there about running examples. Maybe it's magic behind `Command::cargo_bin`.
4. Fails to find the executable at runtime. Okay, let's actually dig in.
5. [The commit](https://github.com/assert-rs/assert_cmd/commit/9379147429ff1eb8cb0766c696d1ae6141b66a33) says users should use `escargot`.
6. What's `escargot`? I can't understand from the repo description, `README.md`, or the docs -- it's perhaps too eccentric.
7. Let's go back to the commit diff, and see how the tests for `cargo_example` changed -- maybe I can copy it.
8. It seems to just call binaries. Hm.
9. What's `escargot`? Took a while to understand that it's programmatically building a `cargo` command to run.
10. Okay let's figure out what methods to call.
11. How do I set environmental variables, or args? In `assert_cmd` I needed the `Command` type.
12. How do I get the command? Oh I have to `CargoBuild::run()` (which doesn't actually *run* the command, haha)
13. Okay now how do I assert that it went happily? Oh I remember `assert_cmd` has a number of traits.

.. later

14. Push to CI, why is this taking so long? (80 minutes, CI is normally done in 25)
15. Oh, it's because I didn't understand `current_target()` and `current_release()`.

Tada 😄!

**Comment by epage** [comment]
Yeah, I did a bad job with the changelog.  I assumed a lot of people were already using `escargot`.

I did document some of this in the permanent documentation: https://docs.rs/assert_cmd/0.11.0/assert_cmd/cargo/index.html

**Comment by epage** [comment]
btw now that you've taken more time with this, besides the problems with the transition, how is this changing working out for you?

**Comment by azriel91** [comment]
The change is good 👌 -- needed that extra tip about `assert()` to shorten the chain.

**Comment by epage** [comment]
I'll at least update the changelog to point to the relevant docs section.
